### DISCUSSION TITLE
How to start a javascript file from this plugin?

### DISCUSSION THREAD

**Discussion by ReaderGuy42** [Taitava/obsidian-shellcommands#308] [Q&A] [answered]
I have a javascript file that counts words in quotes. According to various threads I should be able to launch a `.js` file from the command line with `node file.js`, but when I try this in the plugin, as a new shell command, the pop up says that `/bin/bash: node: command not found`.

How can I launch a .js file from this plugin? This is on Linux Mint, btw.

Thanks!

It probably helps if `nodejs` is installed on the system haha. But now I'm getting another error:
`Reference error: document is not defined...`

This is the script: 

https://forum.obsidian.md/t/best-way-to-count-words-in-quotes/50507/7?u=readerguy42

It counts the words in quotes and pops out a little window with that info. Is this possible to do via the plugin?

**Comment by Taitava** [comment, marked as answer]
Hi,

thank you for your question. I took a look at the script you are trying to evaluate with Node.js.

```js
// get all cm-lines in active tab
var lineText = document.querySelectorAll(".workspace-leaf.mod-active .cm-content .cm-line");
```

This code should be evaluated in Obsidian app, but your approach to launch a separate Node.js process would evaluate the JS code outside Obsidian. It seems that Node.js does not have a `document` object itself, which is why you are getting the error message `Reference error: document is not defined...`. But even if `document` would be present, the code would still not work outside Obsidian, because it would not have access to the content structure of Obsidian.

If you want to use the code, it's better to use the [JavaScript Init](https://obsidian.md/plugins?id=obsidian-javascript-init) plugin like was suggested in the forum thread you linked. Then the code will run in the context of Obsidian.

---

The same behavior could be achieved with SC, too, but the script would need to be edited so that it would not use `document` at all, but it would rather take a file path as a parameter and read the content from the file. But as it looks like you already have script that should work using the other plugin, I guess it's easier to use that plugin instead of trying to modify the script to work outside Obsidian.

**Reply by ReaderGuy42** [reply]
All right, thank you :)

**Comment by Taitava** [comment]
# Concerning your other question in #306

Quote from @ReaderGuy42 [in #306](https://github.com/Taitava/obsidian-shellcommands/issues/306#issuecomment-1367815362):
> Hi, I have a related question, I think. I'm trying to run
> `xdotool key ctrl+a & xdotool key ctrl+c & xdotool key ctrl+g`, but I don't think it's working. Ctrl+g here is the quote word count that I was trying to get going the other day, and in order to count ALL words in a longer note it needs to be cached or something via Ctrl+A (select all), Ctrl+C (Copy).
> 
> If I just run Ctrl+g in ShellCommands, the pop up from this custom quote counter plugin comes up. But Ctrl+A on its own does not work. Just trying to streamline this process.
>
> Any ideas? Thanks :)

Sorry, I don't know `xdotool` at all. Seems like you got automatic hotkey pressing working almost well, with the exception that Ctrl + A does not work. I've never tried to make SC press hotkeys myself, but if I remember correctly @FelipeRearden has posted some examples or questions about shell commands that press hotkeys, but that was on macOS if I remember correctly, while you are using Linux. I can't remember the command Felipe used and so cannot tell does it work on Linux. Maybe I'll find it some day somewhere from the Discussions here. Sorry, I don't have time to look for it right now.

Even if I find it, it might be that it does not solve your problem, but could at least try.

**Reply by ReaderGuy42** [reply]
OK, thanks for the effort! Not a rush :)
Thanks!

**Reply by Taitava** [reply]
Sorry, it took some time for me to come back to this. I tried to look for an alternative command that would press hotkeys, but didn't find. Maybe I remebered incorrectly that there was that kind of command mentioned in some discussion of this repo.

This is just some blind guessing:

- Have you tried to change `&` to `&&`: `xdotool key ctrl+a && xdotool key ctrl+c && xdotool key ctrl+g`
- Or make it multiline:
  ```
  xdotool key ctrl+a
  xdotool key ctrl+c 
  xdotool key ctrl+g
  ```
- How about if you add some delays:
  ```
  xdotool key ctrl+a
  sleep 1
  xdotool key ctrl+c 
  sleep 1
  xdotool key ctrl+g
  ```
  So that the ctrl+a (select all) command has enough time to perform its action before the next command is called.

---

Finally: I think the best overall solution would be to edit the custom plugin you are using, to make it select all text and use that instead of clipboard. 🙂

**Reply by ReaderGuy42** [reply]
None of those work, but I'll look into editing the plugin, thanks anyways :)

**Reply by Taitava** [reply]
Too bad 🙁. I hope you'll be able to find a solution. 👍
